src/section/section_proof_chain.rs

```diff
@@ -39,258 +38,209 @@ impl SectionProofBlock {
+        &self.key_info.prefix
     }
 
     pub fn version(&self) -> u64 {
-        self.key_info.version()
+        self.key_info.version
     }
 }
 
+/// Chain of section BLS keys where every key is proven (signed) by the previous key, except the
+/// first one.
 #[derive(Debug, Eq, PartialEq, Clone, Hash, Serialize, Deserialize)]
-pub struct SectionProofSlice {
-    /// The version of the section key to use as root of trust.
-    version: u64,
-    /// The prefix of section key to use as root of trust.
-    prefix: Prefix<XorName>,
-    /// chain of trust to the section, if empty use root of trust.
-    blocks: Vec<SectionProofBlock>,
+pub struct SectionProofChain {
+    head: SectionKeyInfo,
+    tail: Vec<SectionProofBlock>,
 }
 
-impl SectionProofSlice {
-    #[cfg(any(feature = "mock_base", test))]
-    pub fn from_genesis(key_info: SectionKeyInfo) -> Self {
+impl SectionProofChain {
+    /// Creates new chain consisting of only one block.
+    pub fn new(first: SectionKeyInfo) -> Self {
         Self {
-            version: key_info.version,
-            prefix: key_info.prefix,
-            blocks: Vec::new(),
+            head: first,
+            tail: Vec::new(),
         }
     }
 
-    pub fn last_prefix_version(&self) -> (&Prefix<XorName>, u64) {
-        self.blocks
-            .last()
-            .map(|block| (block.prefix(), block.version()))
-            .unwrap_or((&self.prefix, self.version))
+    pub(crate) fn push(&mut self, block: SectionProofBlock) {
+        if !validate_next_block(self.last_key_info(), &block) {
+            log_or_panic!(
+                log::Level::Error,
+                "Invalid next block: {:?} -> {:?}",
+                self.last_key_info(),
+                block
+            );
+            return;
+        }
+
+        self.tail.push(block)
     }
 
-    #[cfg(all(test, feature = "mock"))]
-    pub fn all_prefix_version(&self) -> impl DoubleEndedIterator<Item = (&Prefix<XorName>, u64)> {
-        iter::once((&self.prefix, self.version)).chain(
-            self.blocks
-                .iter()
-                .map(|block| (block.prefix(), block.version())),
-        )
+    #[cfg(test)]
+    pub fn self_validate(&self) -> bool {
+        let mut current = &self.head;
+        for block in &self.tail {
+            if !validate_next_block(current, block) {
+                return false;
+            }
+
+            current = block.key_info();
+        }
+        true
     }
 
-    pub fn last_key_info(&self) -> Option<&SectionKeyInfo> {
-        self.blocks.last().map(|block| block.key_info())
+    pub(crate) fn first_key_info(&self) -> &SectionKeyInfo {
+        &self.head
     }
 
-    fn last_trusted_key_info<'a>(
-        &'a self,
-        last_trusted: &'a SectionKeyInfo,
-    ) -> Option<&'a SectionKeyInfo> {
-        let block_offset = last_trusted.version().saturating_sub(self.version) as usize;
+    pub(crate) fn last_key_info(&self) -> &SectionKeyInfo {
```

Comment:
when tail is empty, this function shall return the head ?
as the comment of `new`  function says `Creates new chain consisting of only one block`, where the first is set to head.

Reply:
Yes, the `head`/`tail` thing is an implementation detail to make it statically checked that the chain always contains at least one block. So if the chain contains only one block, it's both the first and the last block.